kernel/modules: kmod-sched-*: fix dependency, add missing modules, support IPSET and BPF

guidosarducci opened the pull request:
**Maintainers/Distribution List**: @ldir-EDB0 
**Compile tested**: mips32-be-malta, openwrt-18.06 & master branch
**Run tested**: mips32-be-malta, openwrt-18.06 & master branch
**Related PRs**: https://github.com/openwrt/openwrt/pull/1615 (merged), https://github.com/openwrt/openwrt/pull/1627

**Description**:
This PR captures a set of related changes to the `NET_SCHED` class of kernel modules defined in `netsupport.mk`, which are configured using the `tc` utility from `iproute2`. The end goal is to enable powerful mangling and filtering capabilities, including using IPSET and BPF.

1. Fix a dependency needed for tc ematch modules, add useful matchall module.
2. Add some common, requested modules currently missing, supporting e.g. packet mangling and logging.
3. Package the tc ematch module supporting classification by IPSET.
4. Package tc classifier and actions support of cBPF and eBPF programs.
5. Package the standard Linux BPF test/verification module, useful to validate BPF functionality or for JIT test/dev.
6. Enable the BPF JIT compiler by default in the 4.9/4.14/4.19 generic kernel configs.

**Sizing Considerations**:
Changes to package and kernel sizes are minimal.

1. New packages for IPSET and BPF support are standalone without impacting existing installs. For reference, package sizes (in bytes) are:
   - `kmod-sched-ipset_4.14.89-1_mips_24kc.ipk`: 2228
   - `kmod-sched-bpf_4.14.89-1_mips_24kc.ipk`: 7536
   - `kmod-bpf-test_4.14.89-1_mips_24kc.ipk`: 18928
2. Minimal impact to kernel size. Enabling BPF JIT increased size of uImage-lzma (normally ~2 MB) by only 5 KB for a cBPF JIT on the MIPS32 arch, and 9 KB for a eBPF JIT on the MIPS64 arch.
3. Slight size increase to existing `kmod-sched*` packages:
   - `kmod-sched-core_4.14.89-1_mips_24kc.ipk`: 49988 (+2.9 KB change)
   - `kmod-sched_4.14.89-1_mips_24kc.ipk`: 52256 (+6.5 KB change)

Appreciate any feedback!

guidosarducci committed:
kernel/modules: add kmod-sched-bpf package

Add cls_bpf and act_bpf modules for additional tc classifier and action
support of cBPF and eBPF.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel/modules: kmod-sched: add some common, useful actions

Add act_pedit, act_csum, act_gact and act_simple modules for additional
tc action support. Module act_simple helps with debug and logging, similar
to iptables LOG target, while act_gact provides common generic actions.
Modules act_pedit and act_csum support general packet mangling, and have
been the subject of feature requests and forum discussions (e.g. DSCP),
as well as being added to the Turris OS fork of OpenWrt ~2 years ago.

Also select dependency kmod-lib-crc32c to support act_csum.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel/modules: kmod-sched-core: add missing dependency, useful module

All tc ematch modules, including those in kmod-sched-core and kmod-sched,
use cls_basic as a core dependency. Relocate cls_basic from kmod-sched to
kmod-sched-core to avoid requiring kmod-sched unnecessarily.

This change is also backwards compatible since any past tc ematch users
will have had to install both kmod-sched-core and kmod-sched anyway.

Add the matchall kernel module cls_matchall introduced in kernel 4.8. The
matchall classifier matches every packet and allows the user to apply
actions on it. It is a simpler, more efficient replacement for the common
but cryptic tc classifier idiom "u32 match u32 0 0".

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel/modules: add kmod-sched-ipset package

Add em_ipset module to support tc filter classification by IP set. Build
as a standalone package to help avoid pulling in rest of kmod-sched and
isolate new dependency on kmod-ipt-ipset.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel/modules: add kmod-bpf-test package

Add the test_bpf module that runs various test vectors against the BPF
interpreter or BPF JIT compiler. The module must be manually loaded, as
with the kmod-crypto-test module which serves a similar purpose.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
kernel: enable CONFIG_BPF_JIT by default

Enable the built-in BPF JIT compiler for all 4.9, 4.14 and 4.19 kernels,
which should speed up cBPF and eBPF-based packet filtering (tc, iptables)
and packet sniffing (libpcap, tcpdump, fwknopd, etc).

This has minimal kernel size impact, increasing the size of uImage-lzma
(normally ~2 MB on mips_24kc or mips64el_mips64) by 5 KB for the MIPS32
arch cBPF JIT and by 9 KB for the MIPS64 arch eBPF JIT, on kernel 4.14.

With JIT enabled (cBPF only), the standard BPF test module (test_bpf.ko)
running on a DIR-835 (mips_24kc) used 33 CPU seconds, but 68 without JIT.

This change aligns with the notion of OpenWRT as the network go-to swiss
army knife for packet handling, especially on CPU-constrained platforms.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
base-files: enable BPF JIT sysctl by default

Set net.core.bpf_jit_enable=1 in /etc/sysctl.d/10-default.conf.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci commented:
I've cleaned up `kmod-sched-bpf` and `kmod-bpf-test` by pulling out enabling of the BPF JIT. I'll rebase on current master first, and then push commits to enable BPF JIT. Hopefully rebasing doesn't confuse Github...

I've also confirmed minimal kernel size impact for the eBPF JIT, and updated the OP as such.

guidosarducci commented:
To whet peoples' appetites, some test cases and examples of using these modules:

- tc ematch ipset (cls_basic, em_ipset)

```
ipset create bulk hash:ip -exist
ipset add bulk 10.20.30.40
tc filter add dev lo parent 1: prio 50 protocol ip basic match 'ipset(bulk dst)' action simple sdata "ipset bulk packet!" continue
ping -c 5 -I lo 10.20.30.40
[327865.135105] simple: ipset bulk packet!_1
[327866.138353] simple: ipset bulk packet!_2
[327867.139261] simple: ipset bulk packet!_3
[327868.140240] simple: ipset bulk packet!_4
[327869.141164] simple: ipset bulk packet!_5
```
- trivial eBPF example from tc-bpf manpage
```
#(compile object on any host)
cat > bpf_test.c <<EOF
#include <linux/bpf.h>

#ifndef __section
# define __section(x)  __attribute__((section(x), used))
#endif

__section("classifier") int cls_main(struct __sk_buff *skb)
{
    return -1;
}

char __license[] __section("license") = "GPL";
EOF
clang-6.0 -O2 -Wall -target bpfeb -c bpf_test.c -o bpf_test.o #(e.g. Malta MIPS BE)

#(copy to openwrt)
scp bpf_test.o root@192.168.1.1:

#(implement on router)
tc qdisc add dev lo root handle 1: fq_codel
tc filter add dev lo parent 1: prio 20 bpf object-file bpf_test.o  classid 1:1  action continue
```
- zero DSCP for IPv4/IPv6 using `act_pedit`
```
tc filter add dev lo parent 1: prio 70 protocol ip u32 match u32 0 0 action pedit munge offset 0 u32 set 0 retain 0xff03ffff continue
tc filter add dev lo parent 1: prio 80 protocol ipv6 u32 match u32 0 0 action pedit munge offset 0 u32 set 0 retain 0xf03fffff continue
```

All these examples have worked for many years on mainstream Linux distros. How great would it be to have them on a dedicated networking platform like OpenWRT?

tohojo commented:
+1 on enabling the bpf jit. Enabling XDP support on some platforms is another case where it can be useful; there are patches in progress for the mvneta driver for that at least...

guidosarducci commented:
Good point about XDP. With the changes in the related #1627 `ip` will support loading XDP programs. I haven't used XDP yet though, due to lack of a suitable device. But I'm curious about the recent "netdevsim".

BTW congrats on the successful defense.

guidosarducci commented:
Thanks for clarifying. I've added the negative depends along with the module.

I'll try rebasing this on `master`, which will add the kernel 4.19 generic config, and then include that in the BPF JIT enabling commit. Let's see if Github chokes displaying the new commit history...

guidosarducci commented:
Updates made successfully but Github display is unhappy. The PR is ready for review so please look at the [master-update-netsupport](https://github.com/guidosarducci/openwrt/commits/master-update-netsupport) branch directly.

@jow- @ldir-EDB0 @tohojo If any of you would find review easier with a new PR (that displayed changes properly) let me know...

tohojo commented:
LGTM

guidosarducci commented:
I noticed there was also a proposal for adding "tc ... action gact" support on [openwrt-devel](http://lists.infradead.org/pipermail/openwrt-devel/2018-December/014908.html) by @FinboySlick (Jonathan Thibault I think?).  Jonathan, appreciate any feedback or review you might make.

Also pinging @dtaht since I recall him sometimes pining for better BPF support, so may have some interest in both this PR and #1627 . Thanks for any feedback!

hauke reviewed (approved):
LGTM

guidosarducci commented on package/kernel/linux/modules/netsupport.mk:
```diff
@@ -716,15 +716,16 @@ $(eval $(call KernelPackage,mppe))
 define KernelPackage/sched-core
   SUBMENU:=$(NETWORK_SUPPORT_MENU)
   TITLE:=Traffic schedulers
+  DEPENDS:=@!LINUX_3_18 @!LINUX_4_1 @!LINUX_4_4
```
Hmm. I was being cautious to prevent `cls_matchall` usage prior to kernel-4.9. Also, I've been testing on both `openwrt-18.06` and `master`, and wanted to allow for a simple backport.

But I see you're right. After double-checking, it seems using only `@!LINUX_3_18` should work for both `openwrt-18.06` and `master` branches, so I'll drop the `@!LINUX_4_1 @!LINUX_4_4` portion.

guidosarducci replied:
I also realized that since the iproute2 `tc` utility is dependent on `kmod-sched-core`, it won't be build-able for linux-3.18 targets (where `kmod-sched-core` is disabled).

@hauke  Is the above OK given all the `master` targets using 3.18 are marked "source-only" or "broken"?

peterwillcn replied:
Add kernel 4.14 and 4.19

guidosarducci replied:
These kernels are already supported, so please clarify your comment further.

guidosarducci commented:
OK, squashed and pushed an update after dropping any DEPENDS on `@!LINUX_4_1 @!LINUX_4_4`. Also removed an unneeded dependency on `kmod-sched-core` from `kmod-sched-bpf`.

Still looking for advice on the potential problem with [linux-3.18 targets](https://github.com/openwrt/openwrt/pull/1626#discussion_r244917062)...

FinboySlick commented:
I use both BPF and IPSET on a regular basis so these changes are definitely something I am interested in.

As far as gact is concerned, it isn't entirely clear to me which package it should to be a part of.  My initial patch added it to kmod-sched-core but that was refused because of devlink and rdma dependencies.

I alternatively suggested just kmod-sched as it seems much too small to warrant its own package but haven't heard back on it.  The ability to accept, drop or reclassify packets seems pretty basic to me and it is already built with the other 'sched' modules so no kernel size changes.

It really is just a matter of where to put it so if we can figure that out here, I have no issue with you including it as part of your changes.

guidosarducci commented:
> I use both BPF and IPSET on a regular basis so these changes are definitely something I am interested in.

Excellent! Always good to find more people actively using BPF/IPSET. How are you using it? Have you tried this PR together with #1627 ?

>As far as gact is concerned, it isn't entirely clear to me which package it should to be a part of.

While I agree that the "ability to accept, drop or reclassify packets seems pretty basic" and module `act_gact` probably belongs in `kmod-sched-core`, I've left it in `kmod-sched` as the least contentious path to make progress. Also, since `kmod-sched` depends on `kmod-sched-core`, it would be simple to move gact to the latter in a safe, backwards compatible way in the future.

>My initial patch added it to kmod-sched-core but that was refused because of devlink and rdma dependencies.

The refusal/objection as I read it was that iproute2/tc and qos-scripts depend on `kmod-sched-core`, and so it shouldn't be changed, but this doesn't make much sense in any light. Packages `tc` and `kmod-sched-core` are dependent by design and together provide networking infrastructure for other packages. One could equally argue that `iptables-mod-ipopt` and `kmod-ipt-ipopt` are dependencies of `iptables`, and thus shouldn't be updated either! It should also be clear the reason for `kmod-sched-core` existence goes beyond just `qos-scripts`. There are and will be other packages that depend on qdiscs, classifiers and actions, some contributed (like `sqm-scripts`) but many others developed and used without being submitted to openwrt repos. 

This PR already has some "LGTM"s from @hauke and @tohojo so hopefully can be merged soon. If they asked to move "gact" into "kmod-sched-core" that would be fine, but I'm honestly just trying to avoid churn at this stage.

Thanks for the feedback! All the best in the new year...

FinboySlick commented:
I manage a WISP and the IPSET is used in our traffic shaper.  BPF is something I am experimenting with as part of a specialized network access platform (which I hope will eventually run on OpenWRT).  I found tc BPF's ability to use the program return value as a target classid to be incredibly useful.

My understanding of why kmod-sched-core should grow as little as possible is that OpenWRT's qos-scripts is very good and effort should be made to keep it as small as possible and available on space-constrained devices.  As such, I have no issue with my initial patch being rejected.  My entire focus is having those features available without having to maintain one's own releases of OpenWRT so whatever gets them in is great from my perspective.

As such, I highly commend your efforts to bring more advanced networking features to the distro. :thumbsup:

guidosarducci commented:
Could we merge this now or is something missing but unspoken?

There's been general approval and desire for making these changes, so pinging relevant committers and those related/interested @hauke @ldir-EDB0 @jow- @tohojo . Thanks for your consideration...

ldir-EDB0 commented:
@guidosarducci  I'll poke this tomorrow and commit it unless someone says no - it's hung around for too long - I suspect it's not been committed 'cos people don't want to break the build

guidosarducci commented:
>I'll poke this tomorrow and commit it unless someone says no - it's hung around for too long - I suspect it's not been committed 'cos people don't want to break the build

@ldir-EDB0 Thanks! I expect the build should be OK since I've been regularly rebasing and rebuilding this (together with #1627) against the master branch and then rerunning some functional tests. It's definitely hung around too long ("why?" is another conversation) so any help finalizing this is appreciated.

For my own future reference, is there a simple way to test building against all supported architectures? Would someone pulling this into their staging tree trigger that? Anything else besides setting up my own buildbot?

ldir-EDB0 commented:
Merged into master!

ldir-EDB0 commented:
@guidosarducci  Pulling stuff into a staging tree doesn't magically mean...anything at all really, just pulled into a staging tree and whatever tests (or none at all) before hopefully being pushed into master.  I'm also not aware of magic invocation to build on >1 target for testing purposes...and in my case limited cpu power makes that challenging.  AFAIK there are plans to enable a buildbot to look at staging trees for that sort of testing purpose to avoid unexpected surprises.

guidosarducci commented:
@ldir-EDB0  It was just a thought about sometimes wanting more coverage exactly "for that sort of testing purpose to avoid unexpected surprises". Manually testing more than 2 archs doesn't scale well, but OTOH doing multi-arch builds per commit on each staging tree seems too resource hungry. I was thinking more along the lines of allowing core developers to make infrequent "spot-requests" to build-bot build a given commit, not necessarily their own staging tree, but e.g. anything late-stage that could benefit from better coverage. Cheers!
